### PR TITLE
[SPARK-32249][INFRA][3.0] Run Github Actions builds in branch-3.0

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to backport the following JIRAs:

- SPARK-32245
- SPARK-32292
- SPARK-32252
- SPARK-32316
- SPARK-32408
- SPARK-32303
- SPARK-32363
- SPARK-32419
- SPARK-32422
- SPARK-32491
- SPARK-32493
- SPARK-32496
- SPARK-32497
- SPARK-32357
- SPARK-32606
- SPARK-32605
- SPARK-32248
- SPARK-32645
- Minor renaming d0dfe49#diff-02d9c370a663741451423342d5869b21

in order to enable GitHub Actions in branch-3.0.

### Why are the changes needed?

To be able to run the tests in branch-3.0. Jenkins jobs are unstable.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

Build in this PR will test.

### CODE REVIEW

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -9,148 +9,242 @@ on:
...
+            network-common, network-shuffle, repl, launcher,
+            examples, sketch, graphx
+          - >-
+            catalyst, hive-thriftserver
+          - >-
+            streaming, sql-kafka-0-10, streaming-kafka-0-10,
+            mllib-local, mllib,
+            yarn, mesos, kubernetes, hadoop-cloud, spark-ganglia-lgpl
+          - >-
+            pyspark-sql, pyspark-mllib
```
Comment (HyukjinKwon) [inline .github/workflows/master.yml:40]: `pyspark-resource` modules does not exist in Spark 3.0. It was added in SPARK-29641.

File: .github/workflows/master.yml [hunk 2]
```diff
@@ -9,148 +9,242 @@ on:
...
-
-  lintr:
-    runs-on: ubuntu-latest
-    name: Linter (R)
-    steps:
-    - uses: actions/checkout@master
-    - uses: actions/setup-java@v1
+        python-version: pypy3
+        architecture: x64
+    - name: Install Python 2.7
```
Comment (HyukjinKwon) [inline .github/workflows/master.yml:126]: Spark 3.0 did not drop Python 2 yet. In master, we test Python 3.6 and 3.8. In branch-3.0, we test Python 3.8 and 2.7.

**Comment by HyukjinKwon** [comment]
Otherwise, it's a clean porting back. I will pick and merge each commit once the tests pass. cc @dongjoon-hyun, @viirya, @gengliangwang.

cc @ScrapCodes and @zhengruifeng FYI. This is a huge PR but none of them are in main codes. Should be okay.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
All tests passed (I am re-running to verify flakiness) except two tests:

-  `ThriftServerQueryTestSuite`:

```
[info] org.apache.spark.sql.hive.thriftserver.ThriftServerQueryTestSuite *** ABORTED *** (2 minutes, 39 seconds)
[info]   The code passed to eventually never returned normally. Attempted 40 times over 30.012199178 seconds. Last failure message: requirement failed: Failed to bind an actual port for HiveThriftServer2. (SharedThriftServer.scala:87)
[info]   org.scalatest.exceptions.TestFailedDueToTimeoutException:
[info]   at org.scalatest.concurrent.Eventually.tryTryAgain$1(Eventually.scala:432)
[info]   at org.scalatest.concurrent.Eventually.eventually(Eventually.scala:439)
[info]   at org.scalatest.concurrent.Eventually.eventually$(Eventually.scala:391)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.eventually(SQLQueryTestSuite.scala:125)
[info]   at org.scalatest.concurrent.Eventually.eventually(Eventually.scala:308)
[info]   at org.scalatest.concurrent.Eventually.eventually$(Eventually.scala:307)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.eventually(SQLQueryTestSuite.scala:125)
[info]   at org.apache.spark.sql.hive.thriftserver.SharedThriftServer.startThriftServer(SharedThriftServer.scala:87)
[info]   at org.apache.spark.sql.hive.thriftserver.SharedThriftServer.$anonfun$beforeAll$4(SharedThriftServer.scala:40)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at scala.util.Try$.apply(Try.scala:213)
[info]   at org.apache.spark.sql.hive.thriftserver.SharedThriftServer.$anonfun$beforeAll$3(SharedThriftServer.scala:40)
[info]   at scala.util.Failure.orElse(Try.scala:224)
[info]   at org.apache.spark.sql.hive.thriftserver.SharedThriftServer.$anonfun$beforeAll$2(SharedThriftServer.scala:40)
[info]   at org.apache.spark.sql.hive.thriftserver.SharedThriftServer.$anonfun$beforeAll$2$adapted(SharedThriftServer.scala:39)
[info]   at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:162)
[info]   at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:162)
[info]   at scala.collection.immutable.Range.foreach(Range.scala:158)
[info]   at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:162)
[info]   at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:160)
[info]   at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:108)
```

Arrow related SparkR tests which will be fixed at https://github.com/apache/spark/pull/29462.

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -9,148 +9,267 @@ on:
...
-        - java: '11'
-          hive: 'hive-1.2'
-        - hadoop: 'hadoop-3.2'
-          hive: 'hive-1.2'
-    name: Build Spark - JDK${{ matrix.java }}/${{ matrix.hadoop }}/${{ matrix.hive }}
-
+        java:
+          - 1.8
+        hadoop:
+          - hadoop2.7
```
Comment (HyukjinKwon) [inline .github/workflows/master.yml:22]: Hadoop is 2.7 by default in 3.0.

**Comment by HyukjinKwon** [comment]
I will port these back once Jenkins tests pass.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: common/tags/src/test/java/org/apache/spark/tags/GitHubActionsUnstableTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@TagAnnotation
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD, ElementType.TYPE})
+public @interface GitHubActionsUnstableTest { }
```
Comment (HyukjinKwon) [inline common/tags/src/test/java/org/apache/spark/tags/GitHubActionsUnstableTest.java:30]: I temporarily added this tag to exclude some tests being failed in GitHub Actions. We should ideally remove this tag and reenable the tests skipped.

FWIW, the failed tests are:
- `ThriftServerQueryTestSuite`
-  `ThriftServerWithSparkContextSuite`
- `VersionSuite`

these do not happen in the master branch, and Jenkins still tests them so would not be a big deal.

`ThriftServerQueryTestSuite` and `ThriftServerWithSparkContextSuite` were fixed, IIRC, with an invasive fix that only landed to the master.

`VersionSuite` failure, I remember this but I could not figure out the root cause at that time.

**Comment by HyukjinKwon** [comment]
Now it should pass all tests. I will port back once I get a green. (except SparkR tests fixed at #29462).

File: common/tags/src/test/java/org/apache/spark/tags/GitHubActionsUnstableTest.java (hunk 4 above)
Comment (HyukjinKwon) [reply common/tags/src/test/java/org/apache/spark/tags/GitHubActionsUnstableTest.java:30]: cc @dongjoon-hyun FYI

**Comment by dongjoon-hyun** [comment]
This goes as a single commit. Did I understand correctly?

**Comment by HyukjinKwon** [comment]
Oh no. I will add one by one. I intentionally made corresponding changes to each appropriate commit. Does it make sense?
If you prefer single commit way, I don't mind. I just thought pushing it as a single commit is less manageable so took this path but maybe too much in away.

**Comment by HyukjinKwon** [comment]
I just turned this as single PR.

**Comment by HyukjinKwon** [comment]
Should be ready to be reviewed or go merging.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you so much always.
Yep. We can proceed R testing separately.

**Comment by dongjoon-hyun** [comment]
Merged to branch-3.0.

**Comment by HyukjinKwon** [comment]
Thanks @dongjoon-hyun!

**Pull request closed without merging**